modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/pagemem/SpeedBasedCleanPagesProtectionThrottle.java

```diff
@@ -0,0 +1,500 @@
+     * store). This ratio is excluded from throttling.
+     */
+    private volatile double initDirtyRatioAtCpBegin = MIN_RATIO_NO_THROTTLE;
+
+    /**
+     * Average checkpoint write speed. Only the current value is used. Pages/second.
+     */
+    private final ProgressInstantSpeedCalculation cpWriteSpeed = new ProgressInstantSpeedCalculation();
+
+    /**
+     * Used for calculating speed of marking pages dirty.
+     * Value from past 750-1000 millis only.
+     * {@link IntervalBasedMeasurement#getSpeedOpsPerSec(long)} returns pages marked/second.
+     * {@link IntervalBasedMeasurement#getAverage()} returns average throttle time.
+     */
+    private final IntervalBasedMeasurement markSpeedAndAvgParkTime;
+
+    /***/
+    SpeedBasedCleanPagesProtectionThrottle(PageMemoryImpl pageMemory,
+                                           IgniteOutClosure<CheckpointProgress> cpProgress,
+                                           IntervalBasedMeasurement markSpeedAndAvgParkTime) {
+        this.pageMemory = pageMemory;
+        this.cpProgress = cpProgress;
+        this.markSpeedAndAvgParkTime = markSpeedAndAvgParkTime;
+
+        totalPages = pageMemory.totalPages();
+    }
+
+    /**
+     * Computes next duration (in nanos) to throttle a thread.
+     * Might return #NO_THROTTLING_MARKER as a marker that no throttling should be applied.
+     *
+     * @return park time in nanos or #NO_THROTTLING_MARKER if no throttling is needed
+     */
+    long protectionParkTime(long curNanoTime) {
+        CheckpointProgress progress = cpProgress.apply();
+        AtomicInteger writtenPagesCounter = progress == null ? null : progress.writtenPagesCounter();
+
+        boolean checkpointProgressIsNotYetReported = writtenPagesCounter == null;
+        if (checkpointProgressIsNotYetReported) {
+            resetStatistics();
+
+            return PagesWriteSpeedBasedThrottle.NO_THROTTLING_MARKER;
+        }
+
+        threadIds.add(Thread.currentThread().getId());
+
+        return computeParkTime(writtenPagesCounter, curNanoTime);
+    }
+
+    /***/
+    private void resetStatistics() {
+        speedForMarkAll = 0;
+        targetDirtyRatio = -1;
+        currDirtyRatio = -1;
+    }
+
+    /***/
+    private long computeParkTime(@NotNull AtomicInteger writtenPagesCounter, long curNanoTime) {
+        final int cpWrittenPages = writtenPagesCounter.get();
+        final long donePages = cpDonePagesEstimation(cpWrittenPages);
+
+        final long markDirtySpeed = markSpeedAndAvgParkTime.getSpeedOpsPerSec(curNanoTime);
+        cpWriteSpeed.setProgress(donePages, curNanoTime);
+        final long curCpWriteSpeed = cpWriteSpeed.getOpsPerSecond(curNanoTime);
+
+        final int cpTotalPages = cpTotalPages();
+
+        if (cpTotalPages == 0) {
+            // ??? It would be great to ask dspavlov (original author) or Alexey Goncharuk (reviewer) why we need
```

Comment:
I guess you'd better not forget to ask them)

Reply:
I think I managed to understand when and why it's needed. I've added some comments and renamed a method to put this knowledge to the code.